### ISSUE TITLE
TooManyPublicMethods in PHPUnit test class

### ISSUE THREAD

**Issue by Torniojaws** [phpmd/phpmd#503]
I saw that ``get|set`` are ignored by default, but is there a reason why eg. PHPUnit tests (= methods starting with ``test``, eg. ``public function testMyFeature``) are not ignored? To get good coverage, there is a high chance that there will be 10+ methods in the unit test (or?).

You can of course create a custom ruleset, but I think this would make sense as a default?

**Comment by ravage84** [comment]
The only argument I see right now is that too many test methods in one test class _could_ be seen as a mess by itself, though I guess anybody with a decently sized test suite encounters this problem quite fast...

Also, too many tests in a test class covering one actual class _could_ be a hint that the class itself does too much or better said is too complex.

I'mt not totally against your proposal...

**Comment by ktomk** [comment]
Well if a class requires that many test methods, this is perhaps a good smell. I'm also not totally against the proposal but how to differ here?  Most likely, the report is useful to change the design. Maybe this limit can be configured?

**Comment by ktomk** [comment]
Perhaps default limit should be seven (7), nine (9) or 12. Perhaps 12 is lax, but nine comes close and most likely is already violated. Perhaps 11 is a good fit for a default value in tests while seven (7) is a good for production? But then some differentiation between test and production need to be made,  Which is 9/10 which is already done.

**Comment by Torniojaws** [comment]
Mostly I'm thinking of the best practice of only one assert per test. The byproduct of this is a large amount of separate tests.

Consider this (silly) example:
```
class Example
{
    public function getAuthorsIDs($databaseInstance, $name)
    {
        $sql = 'SELECT id, author FROM table WHERE author = :name';
        $pdo = array('name' => $name);
        return $databaseInstance->run($sql, $pdo);
    }
}
```
Bad practice:
```
public function testGettingAuthorsIDs()
{
    $name = 'Example'; // Where DB does contain valid results for this name
    $result = $this->example->getAuthorsIDs($this->database, $name);

    // Should only have one assert, according to best practices
    $this->assertTrue(empty($result) == false);
    $this->assertTrue(count($result) > 2); // When test results are known to be 3 or more
    $this->assertEquals('Example', $result[0]['author']);
    $this->assertTrue(is_numeric($result[0]['id']);
}
```
Good practice:
```
public function testGettingAuthorsIDsIsNotEmpty()
{
    $name = 'Example'; // Where DB does contain valid results for this name
    $result = $this->example->getAuthorsIDs($this->database, $name);

    $this->assertTrue(empty($result) == false);
}

public function testGettingAuthorsIDsHasMinimumThreeResults()
{
    $name = 'Example'; // Where DB does contain valid results for this name
    $result = $this->example->getAuthorsIDs($this->database, $name);

    $this->assertTrue(count($result) > 2); // When test results are known to be 3 or more
}

public function testGettingAuthorsIDsHasCorrectName()
{
    $name = 'Example'; // Where DB does contain valid results for this name
    $result = $this->example->getAuthorsIDs($this->database, $name);

    $this->assertEquals('Example', $result[0]['author']);
}

public function testGettingAuthorsIDsIsANumber()
{
    $name = 'Example'; // Where DB does contain valid results for this name
    $result = $this->example->getAuthorsIDs($this->database, $name);

    $this->assertTrue(is_numeric($result[0]['id']);
}
```
So just from a simple function, you already have 4 tests. Not to mention when your class has eg. 5 methods, where the other methods are even more complex than the above.

**Comment by ravage84** [comment]
> Mostly I'm thinking of the best practice of only one assert per test.

@Torniojaws I strongly disagree on this. With your (test) code, the amount of lines and especially the duplicated code explodes for no apparent reason.

I can agree that one shall not put coverage of all aspects of a function into one test but your example is on the opposite extreme of the scale. But again, if your code has so many different aspects, it might be a hint for another problem.

Also, you should consider using the various other useful and more specialized assertion-functions of PHPUnit, other than just ``assertTrue()``.

**Comment by Torniojaws** [comment]
After some further reading on asserts, it appears that the "One assert per test" recommendation is more about one abstract assert, and not literal assert. So, it is okay to have multiple asserts in a single test, as long as they all test the same abstract assert.

```
public class testReturnsValidMultiplicationResult()
{
    // This example would multiple each array item, ie. 2*3*4 = 24
    $result = $this->calculator->multiply(array(2, 3, 4));

    // Multiple asserts are OK, since they test the same abstract assert
    $this->assertTrue(is_numeric($result));
    $this->assertEquals(24, $result);
}
```

Originally I thought the best practice meant that there should be a separate test for the value being an integer, and that the result is correct. So in the above example, there would have been two separate methods.
